src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -91,55 +83,75 @@ pub async fn multi_bit_permutation<'a, F: Field, S: SecretSharing<F>, C: Context
     Ok(one_off_permutation)
 }
 
-/// For a given `idx` check if each of the record has same value as idx.
-/// Steps
-/// 1. Get bit representation of `idx`
-/// 2. Calculate equality check
-///   i. keep record bit if `idx` bit is 1
-///   ii.toggle record bit if idx bit is 0 (Done by taking `share_of_one` - value)
-/// 3. Multiply equality checks for all bits per record - in clear per record this will be 1 only for 1 index while 0 for all others
-async fn get_bit_equality_checkers<F, C, S>(
-    idx: usize,
-    multi_bit_input: &[Vec<S>],
+async fn check_everything<F, C, S>(
     ctx: C,
+    record_idx: usize,
+    input: &[Vec<S>],
 ) -> Result<Vec<S>, Error>
 where
     F: Field,
     C: Context<F, Share = S>,
     S: SecretSharing<F>,
 {
-    let num_multi_bits = multi_bit_input.len();
-    assert_ne!(num_multi_bits, 0);
-    let num_records = multi_bit_input[0].len();
-
-    let ctx_across_bits = ctx.narrow(&MultiplyAcrossBits);
-    let idx_in_bits = &idx.view_bits::<Lsb0>()[..num_multi_bits];
-
-    try_join_all(
-        (0..num_records)
-            .zip(repeat(ctx_across_bits))
-            .map(|(rec, ctx)| async move {
-                let share_of_one = ctx.share_of_one();
-
-                let mult_input = zip(multi_bit_input, idx_in_bits)
-                    .map(|(single_bit_input, bit)| {
-                        if *bit {
-                            single_bit_input[rec].clone()
-                        } else {
-                            -single_bit_input[rec].clone() + &share_of_one
-                        }
-                    })
-                    .collect::<Vec<_>>();
-
-                multiply_all_shares(
-                    ctx,
-                    RecordId::from(idx * num_records + rec),
-                    mult_input.as_slice(),
-                )
-                .await
-            }),
-    )
-    .await
+    let record_id = RecordId::from(record_idx);
+    let num_bits = input.len();
+    let mut precomputed_combinations = Vec::with_capacity(1 << num_bits);
+    precomputed_combinations.push(ctx.share_of_one());
+    #[allow(clippy::needless_range_loop)]
+    for bit_idx in 0..num_bits {
```

Comment:
```suggestion
    for (bit_idx, iter) in input.iter().enumerate() {
        let bit = &iter[record_idx];
```

Reply:
Thank you again. I'm incorporating this as well.